Update CMakeLists and dependencies

JushBJJ opened the pull request:
I added cmake dependencies from TT-Metal so this repo is up to date relative to the latest commit and can work hopefully outside of the box without being too complicated to setup. Includes and link libs are also updated

There are two problems i faced and i wanna know if its the same with you @marty1885 

1. When the compiler specifically specified in the cmake command it starts looping trying to find a C/C++ compiler.
2. Can't compile in gcc (my version was v12)

JushBJJ committed:
Update CMakeList and dependencies

JushBJJ committed:
Use normalized_arch_name instead of arch_name

JushBJJ committed:
Update compile options

JushBJJ committed:
Use target_compile_options

marty1885 commented on CMakeLists.txt:
```diff
@@ -17,45 +19,74 @@ if("$ENV{ARCH_NAME}" STREQUAL "wormhole_b0")
+if(CMAKE_CXX_COMPILER AND CMAKE_C_COMPILER)
+    message(STATUS "Using specifed C++ compiler: ${CMAKE_CXX_COMPILER}")
+    message(STATUS "Using specifed C compiler: ${CMAKE_C_COMPILER}")
+else()
+    message(STATUS "No C or C++ compiler specified, defaulting to Clang-17")
+    FIND_AND_SET_CLANG17()
+endif()
```
Please leave the default to whatever the system uses.

marty1885 commented on CMakeLists.txt:
```diff
@@ -17,45 +19,74 @@ if("$ENV{ARCH_NAME}" STREQUAL "wormhole_b0")
+endif()
+
+list(PREPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
+set(ENV{CPM_SOURCE_CACHE} "${PROJECT_SOURCE_DIR}/.cpmcache")
```
Please respect the user's CPM cache configuration if they already have one.

JushBJJ replied:
What do you mean by this? Like using the CPM cache that's already in their tt-metal repo?

JushBJJ replied:
See https://github.com/tenstorrent/cpp-ttnn-project-template/blob/main/sources/sample_lib/CMakeLists.txt

marty1885 replied:
> Like using the CPM cache that's already in their tt-metal repo

No like `CPM_SOURCE_CACHE` can be set by the user from shell. The code snippet force overwrites it (ignores the user's settings). IMHO Tenstorrent should have respect the user's settings. It's their bug. Let's not replicate it.

marty1885 commented on CMakeLists.txt:
```diff
@@ -17,45 +19,74 @@ if("$ENV{ARCH_NAME}" STREQUAL "wormhole_b0")
+    Boost::container
+    tt_metal
     $ENV{TT_METAL_HOME}/build/lib/_ttnn.so
+    $ENV{TT_METAL_HOME}/build/lib/libdevice.so
+    $ENV{TT_METAL_HOME}/build/lib/libtt_metal.so
```
The paths are covered by `target_link_directories`. The only reason `_ttnn.so` is different is it doesn't follow the standard `libXXXX` conversion which linker assumes. Please revert this change unless there's other effects you are aiming for

JushBJJ replied:
thanks for pointing it out, yes libtt_metal wasnt needed. libdevice is though

marty1885 commented on CMakeLists.txt:
```diff
@@ -17,45 +19,74 @@ if("$ENV{ARCH_NAME}" STREQUAL "wormhole_b0")
+    $ENV{TT_METAL_HOME}/build/lib/libtt_metal.so
 )
 
+target_compile_options(ttnn-hello PRIVATE $<$<COMPILE_LANG_AND_ID:CXX,Clang>:-stdlib=libc++>)
```
I recommend add a `USE_LIBCPP` flag instead for forcing libc++ when clang is detected. Clang works perfectly with stdlibc++ on Arch and Metal built on Arch.

marty1885 commented on CMakeLists.txt:
```diff
@@ -17,45 +19,74 @@ if("$ENV{ARCH_NAME}" STREQUAL "wormhole_b0")
 )
 
+target_compile_options(ttnn-hello PRIVATE -fsized-deallocation)
+if(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+    target_link_options(ttnn-hello PRIVATE -lc++ -lc++abi)
+endif()
```
DITTO

marty1885 commented on CMakeLists.txt:
```diff
@@ -17,45 +19,74 @@ if("$ENV{ARCH_NAME}" STREQUAL "wormhole_b0")
     FMT_HEADER_ONLY
 )
 
+target_compile_options(ttnn-hello PRIVATE -fsized-deallocation)
```
I believe this is not available on clang-17? I might be wrong

marty1885 commented on cmake/CPM.cmake:
```diff
@@ -0,0 +1,27 @@
+    EXPECTED_HASH SHA256=${CPM_HASH_SUM}
+)
+
+set(ENV{CPM_SOURCE_CACHE} "${PROJECT_SOURCE_DIR}/.cpmcache")
+include(${CPM_DOWNLOAD_LOCATION})
```
DITTO. Please respect the user's settings and only have an opinion when the user doesn't have one.

marty1885 reviewed (commented):
Thanks for updating the repo! I've left some comments. Please let me know if there is anything I can help with.

marty1885 commented:
@JushBJJ 


> When the compiler specifically specified in the cmake command it starts looping trying to find a C/C++ compiler.

May you show the symptoms? i believe this is because you try to set compilers in the script and CMake needs a re-run after compiler change.


> Can't compile in gcc (my version was v12)

IIRC the minimal is 13

marty1885 commented:
Ping @JushBJJ

JushBJJ commented:
Getting on it...

JushBJJ committed:
Update CMakeLists.txt

- Add USE_LIBCPP flag
- Use system default compiler if no cmake compiler defined
- Allow user to set cpm source cache
- Add axtra directories required
- Add option for libcpp
- Update dependencies

marty1885 commented on cmake/compilers.cmake:
```diff
@@ -0,0 +1,39 @@
+function(FIND_AND_SET_CLANG17)
+    find_program(CLANGPP_17 clang++-17)
+    find_program(CLANG_17 clang-17)
+
+    if(NOT CLANGPP_17 OR NOT CLANG_17)
+        message(FATAL_ERROR "Clang-17 not found. Make sure you have clang-17 and clang++-17 installed and in your PATH")
+    endif()
+
+    set(CMAKE_CXX_COMPILER "${CLANGPP_17}" PARENT_SCOPE)
+    set(CMAKE_C_COMPILER "${CLANG_17}" PARENT_SCOPE)
+endfunction()
+
+function(CHECK_COMPILERS)
+    message(STATUS "Checking compilers")
+
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+        if(ENABLE_LIBCXX)
+            find_library(LIBC++ c++)
+            find_library(LIBC++ABI c++abi)
+            if(NOT LIBC++ OR NOT LIBC++ABI)
+                message(
+                    FATAL_ERROR
+                    "libc++ or libc++abi not found. Make sure you have libc++ and libc++abi installed and in your PATH"
+                )
+            endif()
+        endif()
+        if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS "17.0.0" OR CMAKE_CXX_COMPILER_VERSION GREATER_EQUAL "18.0.0")
+            message(WARNING "Only Clang-17 is tested right now")
+        endif()
+    elseif(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+        if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS "12.0.0")
+            message(FATAL_ERROR "GCC-12 or higher is required")
+        elseif(CMAKE_CXX_COMPILER_VERSION GREATER_EQUAL "13.0.0")
+            message(WARNING "Only GCC-12 is tested right now")
+        endif()
+    else()
+        message(FATAL_ERROR "Unsupported compiler: ${CMAKE_CXX_COMPILER_ID} ! Only Clang and GCC are supported")
+    endif()
```
Mostly good now! I believe this file is unused. Let's delete it?

JushBJJ committed:
Remove compilers.cmake

JushBJJ commented:
Done, sorry i was in a hurry earlier on cause i had to go out lol.

Does using g++-12 work for you? I get an error:
```
CMake Error at /usr/share/cmake-3.22/Modules/CMakeTestCXXCompiler.cmake:62 (message):
  The C++ compiler

    "/usr/bin/g++-12"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/jush/ttnn-helloworld-cpp/build/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/usr/bin/gmake -f Makefile cmTC_22194/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_22194.dir/build.make CMakeFiles/cmTC_22194.dir/build
    gmake[1]: Entering directory '/home/jush/ttnn-helloworld-cpp/build/CMakeFiles/CMakeTmp'
    Building CXX object CMakeFiles/cmTC_22194.dir/testCXXCompiler.cxx.o
    /usr/bin/g++-12   -stdlib=libc++  -o CMakeFiles/cmTC_22194.dir/testCXXCompiler.cxx.o -c /home/jush/ttnn-helloworld-cpp/build/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
    g++-12: error: unrecognized command-line option ‘-stdlib=libc++’
    gmake[1]: *** [CMakeFiles/cmTC_22194.dir/build.make:78: CMakeFiles/cmTC_22194.dir/testCXXCompiler.cxx.o] Error 1
    gmake[1]: Leaving directory '/home/jush/ttnn-helloworld-cpp/build/CMakeFiles/CMakeTmp'
    gmake: *** [Makefile:127: cmTC_22194/fast] Error 2
```

bit weird

marty1885 commented:
How did you get CMake to use g++-12? The error deesn't feel like a Metalium problem, but something wrong with how CMake is setup.
